pytensor/compile/function/types.py

```diff
@@ -388,31 +388,48 @@ def __init__(
         self.nodes_with_inner_function = []
         self.output_keys = output_keys
 
-        # See if we have any mutable / borrow inputs
-        # TODO: this only need to be set if there is more than one input
-        self._check_for_aliased_inputs = False
-        for i in maker.inputs:
-            # If the input is a shared variable, the memory region is
-            # under PyTensor control and so we don't need to check if it
-            # is aliased as we never do that.
-            if (
-                isinstance(i, In)
-                and not i.shared
-                and (getattr(i, "borrow", False) or getattr(i, "mutable", False))
+        assert len(self.output_storage) == len(self.maker.fgraph.outputs)
+
+        # Group indexes of inputs that are potentially aliased to each other
+        # Note: Historically, we only worried about aliasing inputs if they belonged to the same type,
+        #  even though there could be two distinct types that use the same kinds of underlying objects.
+        potential_aliased_input_groups = []
```

Comment:
Perhaps this logic could be a utility function? It is reused almost identically below

Reply:
What is reused almost identically? Also note that I want to remove this whole logic soon: https://github.com/pymc-devs/pytensor/issues/1026

Reply:
Making groups from a list of elements based on an equivalence function

Reply:
Maybe but the goal of the PR is to reduce the overhead of the `function.__call__` which usually runs in a hot loop. Adding nested function calls may defeat the purpose. Perhaps I will just go ahead with #1026 and shed this whole logic.

I agree it's ugly to repeat the code